### PR TITLE
PRO and AUPRO metrics

### PR DESCRIPTION
**Pull request by djdameln**
# Description

- This PR adds the PRO AUPRO metrics, and the metrics are used by the base task. The PRO score is evaluated at the optimal threshold (based on F1).

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Known Issues
- The PRO metric has a high computational complexity. The current implementation is optimized to run on the GPU, running it on CPU will cause significant delays in the code execution.

# Checklist:

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/utils/metrics/aupro.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        comps, n_comps = connected_components(targets.unsqueeze(1))
+        self.comps.append(comps.to(predictions.device))
+        self.n_regions.append(torch.tensor(n_comps - 1))  # pylint: disable=not-callable
+
+    def compute(self) -> Tensor:
+        """Compute the macro average of the PRO score across all regions in all batches."""
+        fpr, _, thresholds = roc(torch.vstack(self.predictions).flatten(), torch.vstack(self.targets).flatten())
+
+        # downsample to speed up computation
+        downsample_idx = torch.linspace(0, len(fpr) - 1, self.auc_samples).long()
```
Comment (ashwinvaidya17) [inline anomalib/utils/metrics/aupro.py:43]: Should we maybe rename it to subsample? Also, is it guaranteed that auc_samples will be less than fpr? The next two lines might throw error for a different value.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks. I have only one question.

**Comment by djdameln** [comment]
This one is outdated. I also have some doubts about the implementation, because it runs too slow on CPU. I will revisit the implementation and submit a new PR. Closing this for now.
